File: fitting data to model/find_model_parameters.m
% find_model_parameters will open data for a given individual under a
% certain test (stimulus like neutral or criticism) and call weightsearch
% inside of fminsearch to find the optimal model parameters to fit the data

function ret=find_model_parameters(individual, test)
% default is first control patient
if nargin < 1, individual = 2303; end
if nargin < 2, test = 'neutral'; end

% these global variables are used in weightsearch to know which data to
% open
global ind typetest;
ind = individual;
typetest = test;

% these global variables will be used in weightsearch for the current
% individual's brain data

% global exec_data salience_data interoceptive_data;

% ------------------------- RECORDED BRAIN DATA ---------------------------
% interpolate the actual brain data so that simulated and actual vectors
% are the same length

% Specify the options for opening all csv data files
opts = delimitedTextImportOptions("NumVariables", 10);
% range and delimiter
opts.DataLines = [1, Inf];
opts.Delimiter = ",";
% column names and types
opts.VariableNames = ["VarName1", "patientID", "time", "condition1", "scanincondition1", "condition2", "condition3", "ignore", "brain"];
opts.VariableTypes = ["double", "double", "double", "double", "double", "double", "double", "double", "double"];
% file level properties
opts.ExtraColumnsRule = "ignore";
opts.EmptyLineRule = "read";

% make sure to change directory to be inside /interoception-modeling

% we must select the correct data to open, based on the global variables
% ind and typetest
if strcmp(typetest, 'neutral') % neutral case
    csv_executive = sprintf('%d_executive_%s_76_scans.csv',ind,typetest);
    filename_executive = fullfile('.','data','neutral_rest_76_scans',csv_executive);
    
    csv_salience = sprintf('%d_salience_forward_%s_76_scans.csv',ind,typetest);
    filename_salience = fullfile('data','neutral_rest_76_scans',csv_salience);
    
    csv_interoceptive = sprintf('%d_interoceptive_forward_%s_76_scans.csv',ind,typetest);
    filename_interoceptive = fullfile('data','neutral_rest_76_scans',csv_interoceptive);
else % criticism case
    csv_executive = sprintf('%d_executive_%s_76_scans.csv',ind,typetest);
    filename_executive = fullfile('.','data','criticism_rest_76_scans',csv_executive);
    
    csv_salience = sprintf('%d_salience_forward_%s_76_scans.csv',ind,typetest);
    filename_salience = fullfile('.','data','criticism_rest_76_scans',csv_salience);
    
    csv_interoceptive = sprintf('%d_interoceptive_forward_%s_76_scans.csv',ind,typetest);
    filename_interoceptive = fullfile('data','criticism_rest_76_scans',csv_interoceptive);
end

% Import the data, rmmissing() removes the row of nan's that comes in the
% data
executiverestcontrols = rmmissing(readtable(filename_executive, opts));
saliencerestcontrols = rmmissing(readtable(filename_salience, opts));
interoceptiverestcontrols = rmmissing(readtable(filename_interoceptive, opts));

% examine this data later... some values are negative even though there are
% no negative values in the simulated data. May need to address this
exec_data = (executiverestcontrols.('brain')-10000)./1000 + 1;
salience_data = (saliencerestcontrols.('brain')-10000)./1000 + 1;
interoceptive_data = (interoceptiverestcontrols.('brain')-10000)./1000 + 1;

plot([exec_data salience_data interoceptive_data])

% start Powell's method with the "well regulated" model
owmat=[  .9        .15        0       0   ;   % external threat
   	     0         .9      .25      .25   ;   % vigilance/salience       % FROM THESE NETWORKS
	  -.25       -.04       .9    -.1   ;   % avoidance/control
         0        .25       .15     .9  ];   % interoception 
     
% initialize correlations and timeseries data in outer function, since they
% will be returned in a struct later
exec_corr=0;
sal_corr=0;
int_corr=0;

convolved_salience_data = zeros(1,60000);
convolved_exec_data = zeros(1,60000);
convolved_interoceptive_data = zeros(1,60000);

interpolated_exec_data = zeros(1,60000);
interpolated_salience_data = zeros(1,60000);
interpolated_interoceptive_data = zeros(1,60000);

     function ret=weightsearch(weights)
         global gwmat gstartstate ginstates;
         
         % where we start -- for now assume all networks start at 0.5
         gstartstate=[0  0.5  0.5  0.5 ]; % threat absent vigilant int
         
         % what comes in
         if strcmp(typetest, 'neutral')
             ginstates=[0 0 0 0 ]; % threat absent vigilant int (neutral)
         else
             ginstates=[0.5 0 0 0 ]; % threat present vigilant int (criticism)
         end
         
         % global weight matrix that will be used in dynamic function
         weight_matrix = reshape(weights, 4, 4);
         gwmat=weight_matrix;
         
         % --------------------------- GET SIMULATED DATA -------------------------
         % run model simulation with current weight matrix to get simulated time data for each
         % brain region
         tstats=bada_nn_1999_2('useglobals',0);
         
         model_salience_time_data = tstats.invec(:,2);
         model_exec_time_data = tstats.invec(:,3);
         model_interoceptive_time_data = tstats.invec(:,4);
         
         % --------------------- CONVOLUTION OF SIMULATED DATA --------------------
         % next we perform convolution on simulated data to make it appear like
         % hemodynamic response
         
         % recording length is 120.24 seconds. Simulated data is 60000 samples, so
         % we have 60000/120.24 = 499.002 samples per second
         resamprate=499.002; %what should this number be?
         % get standard hemodynamic response
         hemoir=spm_hrf(1./resamprate, [6,16,1,1,6,0,32]); % start out 100 times per second
         
         % we fill out all convolved data to be 60000 in length, padded with the
         % last value
         convolved_salience_data = conv(model_salience_time_data, hemoir, 'same');
         % convolved_salience_data(length(convolved_salience_data)+1:60000)=convolved_salience_data(length(convolved_salience_data));
         convolved_exec_data = conv(model_exec_time_data, hemoir, 'same');
         convolved_interoceptive_data = conv(model_interoceptive_time_data, hemoir, 'same');
         
         % ------------------------- RECORDED BRAIN DATA ---------------------------
         % interpolate all brain data from glabal vars to be the same scale as simulated data
         % note that resample() will produce edge effects because it assumes there
         % are 0's at the start and end of the data. To account for this, we:
         % 1. subtract mean from original data (data set now has mean of 0)
         % 2. resample the mean normalized data
         % 3. add the mean to the resampled data
         exec_data_mean_norm = exec_data - mean(exec_data);
         interpolated_exec_data = resample(exec_data_mean_norm, length(convolved_exec_data), length(exec_data_mean_norm)) + mean(exec_data);
         sal_data_mean_norm = salience_data - mean(salience_data);
         interpolated_salience_data = resample(sal_data_mean_norm, length(convolved_salience_data), length(sal_data_mean_norm)) + mean(salience_data);
         int_data_mean_norm = interoceptive_data - mean(interoceptive_data);
         interpolated_interoceptive_data = resample(int_data_mean_norm, length(convolved_salience_data), length(int_data_mean_norm)) + mean(interoceptive_data);
         
         plot ([interpolated_exec_data interpolated_salience_data interpolated_interoceptive_data convolved_exec_data convolved_salience_data convolved_interoceptive_data])
         legend({'fMRI exec', 'fMRI salience', 'fMRI int', 'simulated exec', 'simulated salience', 'simulated int'})
         drawnow;
         
         exec_corr = r(interpolated_exec_data, convolved_exec_data);
         sal_corr = r(interpolated_salience_data, convolved_salience_data);
         int_corr = r(interpolated_interoceptive_data, convolved_interoceptive_data);
         
         % calculate our loss function
         loss = mean(sqrt((convolved_exec_data - interpolated_exec_data).^2 + (convolved_salience_data - interpolated_salience_data).^2 + (convolved_interoceptive_data - interpolated_interoceptive_data).^2));
         if loss < 0.0003, loss = 0; end
         
         ret=loss;
     end

weight = reshape(owmat,1,16);

modelparams=fminsearch(@weightsearch, weight);

%save model parameters, correlations, and timeseries fMRI and simulated
%data to a struct
return_struct.weights = modelparams;
return_struct.exec_corr = exec_corr;
return_struct.sal_corr = sal_corr;
return_struct.int_corr = int_corr;

timeseries.fMRI = [interpolated_salience_data, interpolated_exec_data, interpolated_interoceptive_data];
timeseries.simulation = [convolved_salience_data, convolved_exec_data, convolved_interoceptive_data];
timeseries.labels = {'salience','executive','interoceptive'};
                
return_struct.timeseries = timeseries;

ret = return_struct;
end

File: python/findmodelparameters.py
# this file contains two functions:
# 1. find_parameters() takes in two arguments
#  individual: int. This is the patient ID
#  test: string. This is the stimulus being tested. Can be neutral or criticism
#  RETURN: returns 1x16 numpy array of parameters fit to data, found by weightsearch()
# 
# 2. weightsearch() takes in one argument
#  weights: numpy array. This is a flattened 1x16 array of the 4x4 weight matrix. 
#  These values are being optimized.
#  RETURN: returns loss function used as error for scipy.optimize

import numpy as np
import pandas as pd
from scipy.optimize import minimize, rosen, rosen_der
from dynamicalmodel import interoceptive_update, external_percept_simulation
from spm_hrf import spm_hrf

def weightsearch(weights): # IMPLEMENT THIS
    # global variables needed to run simulated model
    global globalweights
    global globalstartstates
    global globalinstates

    # global variable needed to dtermine if the in state should include 
    global typetest

    # make globalstartstates globalinstates 
    globalstartstates = np.array([[0, 0.5, 0.5, 0.5]]) # all networks start at 0.5 activation
    globalinstates = np.array([[0, 0, 0, 0]]) # this is neutral case
    globalweights = np.reshape(weights, 4, 4)

    simulateddata = external_percept_simulation('useglobals')

    # since we passed in instates with only one row, only concerned with first layer of simulateddata
    # look at weight matrix info in dynamicalmodel to see which rows hold data for each region
    simulatedexecdata = simulateddata[0, 1, :] 
    simulatedsaliencedata = simulateddata[0, 2, :]
    simulatedinteroceptivedata = simulateddata[0, 3, :]

    # convolution with hemodynamic response function to get simulated BOLD signal
    resamprate = 100
    hemoir = smp_hrf(1/resamprate, [6, 16,1, 1, 6, 0, 32])

    convolvedexecdata = np.convolve(simulatedexecdata, hemoir)
    convolvedsaliencedata = np.convolve(simulatedsaliencedata, hemoir)
    convolvedinteroceptivedata = np.convolve(simulatedinteroceptivedata, hemoir)

    # we need to interpolate fMRI data to be the same length as convolved simulated data
     

    return np.mean(weights)

def find_parameters(individual = 2303, test='neutral'):
    # construct name of files that you want to open
    if test=='neutral': # neutral stimulus
        csv_exec = f'../data/neutral_rest_76_scans/{individual}_executive_{test}_76_scans.csv'
        csv_salience = f'../data/neutral_rest_76_scans/{individual}_salience_forward_{test}_76_scans.csv'
        csv_interoceptive = f'../data/neutral_rest_76_scans/{individual}_interoceptive_forward_{test}_76_scans.csv'
    else: # criticism stimulus
        csv_exec = f'../data/criticism_rest_76_scans/{individual}_executive_{test}_76_scans.csv'
        csv_salience = f'../data/criticism_rest_76_scans/{individual}_salience_forward_{test}_76_scans.csv'
        csv_interoceptive = f'../data/criticism_rest_76_scans/{individual}_interoceptive_forward_{test}_76_scans.csv'

    # grab brain data in prefrontal cortex, amygdala, insula. Shift and scale
    exec_brain_data = (np.array(pd.read_csv(csv_exec)['brain'])-10000)/1000+1
    exec_brain_data = (np.array(pd.read_csv(csv_salience)['brain'])-10000)/1000+1
    exec_brain_data = (np.array(pd.read_csv(csv_interoceptive)['brain'])-10000)/1000+1

    starting_point = np.array([[0.9, 0.15, 0, 0],
                               [0, 0.9, 0.25, 0.25],
                               [-.25, -.04, .9, -.1],
                               [0, .25, .15, .9]])
                                
    weights = np.reshape(starting_point, 1, 16)

    optimized_weights = minimize(weightsearch, weights, method='Powell')
    return optimized_weights

